store/src/java/com/zimbra/cs/mailbox/LocalMailboxCacheManager.java

```diff
@@ -0,0 +1,296 @@
+package com.zimbra.cs.mailbox;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.zimbra.common.localconfig.LC;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.db.DbMailbox;
+import com.zimbra.cs.db.DbPool;
+
+import java.lang.ref.SoftReference;
+import java.util.*;
+
+public class LocalMailboxCacheManager implements MailboxCacheManager {
+    /**
+     * Maps account IDs (<code>String</code>s) to mailbox IDs
+     * (<code>Integer</code>s).  <i>Every</i> mailbox in existence on the
+     * server appears in this mapping.
+     */
+    private Map<String, Integer> mailboxIds;
+    /**
+     * Maps mailbox IDs ({@link Integer}s) to either
+     * <ul>
+     * <li>a loaded {@link Mailbox}, or
+     * <li>a {@link SoftReference} to a loaded {@link Mailbox}, or
+     * <li>a {@link MaintenanceContext} for the mailbox.
+     * </ul>
+     * Mailboxes are faulted into memory as needed, but may drop from memory when the SoftReference expires due to
+     * memory pressure combined with a lack of outstanding references to the {@link Mailbox}.  Only one {@link Mailbox}
+     * per user is cached, and only that {@link Mailbox} can process user requests.
+     */
+    private MailboxMap cache;
+
+    public LocalMailboxCacheManager() {
+        DbPool.DbConnection conn = null;
+        synchronized (this) {
+            try {
+                conn = DbPool.getConnection();
+                mailboxIds = DbMailbox.listMailboxes(conn, null);
+                cache = createCache();
+            }catch (ServiceException e) {
+                ZimbraLog.mailbox.error("creating cache manager: ", e);
+                throw new CacheManagerException("creating cache manager: ", e.getCause());
+            } finally {
+                DbPool.quietClose(conn);
+            }
+        }
+    }
+
+    @Override
+    public Collection<Integer> getMailboxIds() {
+        return mailboxIds.values();
+    }
+
+    public Set<String> getAccountIds(){
+        return mailboxIds.keySet();
+    }
+
+    @Override
+    public Set<Map.Entry<String, Integer>> getMailboxIdsByAccountId(){
+        return mailboxIds.entrySet();
+    }
+
+    @Override
+    public int getMailboxCount(){
+        return mailboxIds.size();
+    }
+
+    @Override
+    public Integer getMailboxKey(String accountId){
+        return mailboxIds.get(accountId);
+    }
+
+    @Override
+    public void removeMailboxId(String accountId){
+        mailboxIds.remove(accountId);;
+    }
+
+    public MailboxMap createCache() {
+        return new MailboxMap(LC.zimbra_mailbox_manager_hardref_cache.intValue());
+    }
+
+    @VisibleForTesting
+    public void clearCache() {
+        cache.clear();
+        mailboxIds.clear();
+    }
+
+    public Object retrieveFromCache(int mailboxId, boolean trackGC, MailboxManager mailboxManager) throws MailServiceException {
+        synchronized (mailboxManager) {
```

Comment:
I think that the `MailboxManager` instance that owns this `MailboxCacheManager` should be responsible for synchronization. You should remove the `mailboxManager` parameter from `MailboxCacheManager#retrieveFromCache` definition.

Reply:
Yes, `MailboxManager` is responsible for synchronization https://github.com/Zimbra/zm-mailbox/blob/eb42e6de99b3108de2b84f6a4ffc7decdeeb65d7/store/src/java/com/zimbra/cs/mailbox/MailboxManager.java#L443 , just extracted `retrieveFromCache` trying to not modify the behavior. I noticed this method is synchronized outside and inside in feature/ha  https://github.com/Zimbra/zm-mailbox/blob/62f85f467f4b4e45b71db2cebef2e6708e486942/store/src/java/com/zimbra/cs/mailbox/MailboxManager.java#L577, How about if we define synchronized this method? or just remove the synchronization inside of the method? . Regarding removing `mailboxManager` parameter , it is still needed in `LocalMailboxCacheManager` https://github.com/Zimbra/zm-mailbox/blob/eb42e6de99b3108de2b84f6a4ffc7decdeeb65d7/store/src/java/com/zimbra/cs/mailbox/LocalMailboxCacheManager.java#L94 . Should we refactor in a different way this method ?